File: scala/src/main/scala/com/gildedrose/GildedRose.scala
package com.gildedrose

class GildedRose(val items: Array[Item]) {

  val brie = "Aged Brie"
  val backstagePass = "Backstage passes to a TAFKAL80ETC concert"
  val sulfuras = "Sulfuras, Hand of Ragnaros"

  def updateQuality() {
    for (i <- 0 until items.length) {
      items(i).name match {
        case `sulfuras` =>
        case `backstagePass` => alterBackstagePassQuality(i)

        case _ =>


          if (items(i).name.equals(brie)) {
            increaseQuality(i)
          } else {
            if (items(i).quality > 0) {
              items(i).quality = items(i).quality - 1
            }
          }

          items(i).sellIn = items(i).sellIn - 1

          if (items(i).sellIn < 0) {
            if (!items(i).name.equals(brie)) {
              if (items(i).quality > 0) {
                items(i).quality = items(i).quality - 1
              }
            } else {
              increaseQuality(i)
            }
          }
      }
    }
  }

  private def alterBackstagePassQuality(i: Int) = {
    increaseQuality(i)

    if (items(i).sellIn < 11) {
      increaseQuality(i)

      if (items(i).sellIn < 6) {
        increaseQuality(i)
      }

      if (items(i).sellIn < 1) {
        items(i).quality = 0
      }
      items(i).sellIn = items(i).sellIn - 1
    }
  }

  private def increaseQuality(i: Int) = {
    if (items(i).quality < 50) {
      items(i).quality = items(i).quality + 1
    }
  }
}

File: Kotlin/src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

class GildedRose(val inventory: Array<Item>) {

    companion object {
        const val MAX_QUALITY = 50
        const val TICKET_FIRST_INCREASE_DATE = 10
        const val TICKET_SECOND_INCREASE_DATE = 5
    }

    fun updateQuality() {
        for (item in inventory) {
            item.applyQualityChanges()
            item.reduceSellInDate()
        }
    }

    private fun Item.applyQualityChanges() {
        when (this.name){
            "Backstage passes to a TAFKAL80ETC concert" -> this.applyTicketQualityIncrease()
            "Aged Brie" -> this.increaseQuality()
            "Sulfuras, Hand of Ragnaros" -> this.sellIn ++
            else -> this.reduceQuality()
        }
    }

    private fun Item.applyTicketQualityIncrease() {
        when {
            this.sellIn <= 0 -> { this.quality = 0 }
            this.sellIn <= TICKET_SECOND_INCREASE_DATE -> repeat(3) { this.increaseQuality() }
            this.sellIn <= TICKET_FIRST_INCREASE_DATE -> repeat(2) { this.increaseQuality() }
            else -> this.increaseQuality()
        }
    }

    private fun Item.reduceQuality() {
        if (this.isAboveMinQuality()) {
            this.quality --
            if (this.isPastSellInDate()) {
                this.quality --
            }
        }
    }

    private fun Item.increaseQuality() {
        if (this.isBelowMaxQuality()) {
            this.quality ++
            if (this.isPastSellInDate()) {
                this.quality ++
            }
        }
    }

    private fun Item.isPastSellInDate(): Boolean {
        return sellIn <= 0
    }

    private fun Item.isBelowMaxQuality(): Boolean {
        return quality < MAX_QUALITY
    }

    private fun Item.isAboveMinQuality(): Boolean {
        return quality > 0
    }

    private fun Item.reduceSellInDate() {
        this.sellIn --
    }
}

